### PR TITLE
[+] Change color function [+] Change font function

### PR DESCRIPTION
**Pull request by eevadai**
[+] Change color function [+] Change font function

### CODE REVIEW

**Comment by Lei-Tin** [comment]
I helped you remove .DS_Store from your branch!

**Comment by Lei-Tin** [comment]
# Nice work! 👍 

After playing around with the setting, I found out a few problems
1. image
There is a space here that doesn't really align well? 

2. The settings window pops out on the top left corner of the monitor, maybe we can try to initialize the window at the center  of the screen. 

3. After setting up the preferences we like, the "background color" and "font" re-initializes to the default of the application instead of keeping the settings that we originally had. 

For example, if I changed the background color to "Red" and the font to "Times New Roman". If I would like to edit the settings again, I would probably want to keep the current settings and adjust from the settings. 

4. After closing the settings menu either using the "X", "Confirm", or "Cancel", it seems like the menu bar's setting is still being pressed, hence, I cannot click on settings again. In order to open the settings menu again, I would need to click elsewhere aside from the settings button to reset the state of the settings button, then I can press on it again. 


I will do code review later this week, but maybe @ScottCTD  can also handle that for me. 
PS: Seems like the settings window is not working normally for Scott in his Windows PC. 
image

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return fontIndex;
+    }
+
+    public boolean exec() {
+        this.setTitle("Setting");
+        this.setModal(true);
+        this.setSize(400, 200);
+        this.setResizable(false);
+
+        SwingUtil.centerInOwner(this, this.getOwner());
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/SettingDialog.java:90]: This line should not be there

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 2]
```diff
@@ -26,17 +34,98 @@ private void initMenu() {
...
 
-        this.add(this.settingsMenu);
         this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
 
         this.add(this.helpMenu);
     }
 
+    // Function to set up the setting dialog
+    private int[] setting() {
+        SettingDialog setdlg = new SettingDialog(this);
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:95]: We should then use `SwingUtil.centerInOwner` to display the setting dialog in the center of our main window.

File: src/main/java/billgates/view/gui/CustomTitleBorder.java [hunk 3]
```diff
@@ -5,12 +5,13 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
-public class CustomTitleBorder extends TitledBorder {
-
+public class CustomTitleBorder extends TitledBorder{
     public static final int DEFAULT_BORDER_THICKNESS = 3;
     public static final int DEFAULT_EMPTY_BORDER_THICKNESS = 7;
-    public static final Color DEFAULT_BORDER_TEXT_COLOR = new Color(220, 120, 150);
+    public static final Color borderTextColor = new Color(220, 120, 150);
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/CustomTitleBorder.java:14]: I don't know what's going on here, but why did you change `DEFAULT_BORDER_TEXT_COLOR` to be `borderTextColor`?

File: src/main/java/billgates/view/gui/MainFrame.java [hunk 4]
```diff
@@ -5,7 +5,7 @@
 import javax.swing.border.LineBorder;
 import java.awt.*;
 
-public class MainFrame extends JFrame {
+public class MainFrame extends JFrame{
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/MainFrame.java:8]: I think there should be a space before `{`

File: src/main/java/billgates/view/gui/MainFrame.java [hunk 5]
```diff
@@ -14,8 +14,8 @@ public class MainFrame extends JFrame {
     public static final Color DEFAULT_BACKGROUND_COLOR = new Color(220, 120, 150, 100);
 
     private final JPanel contentPane = new JPanel(new BorderLayout());
-    private final ActionPanel actionPanel = new ActionPanel();
-    private final BillPanel billPanel = new BillPanel();
+    private static final ActionPanel actionPanel = new ActionPanel();
+    private static final BillPanel billPanel = new BillPanel();
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/MainFrame.java:18]: I think `actionPanel` and `billPanel` should not be `static`

File: src/main/java/billgates/view/gui/MainFrame.java [hunk 6]
```diff
@@ -46,11 +46,11 @@ public MainFrame() {
...
     }
 
-    public BillPanel getBillPanel() {
-        return this.billPanel;
+    public static BillPanel getBillPanel() {
+        return billPanel;
     }
 
-    public ActionPanel getActionPanel() {
+    public static ActionPanel getActionPanel() {
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/MainFrame.java:53]: Same problem

File: src/main/java/billgates/view/gui/MainFrame.java [hunk 7]
```diff
@@ -46,11 +46,11 @@ public MainFrame() {
         this.add(this.billPanel, BorderLayout.CENTER);
     }
 
-    public BillPanel getBillPanel() {
-        return this.billPanel;
+    public static BillPanel getBillPanel() {
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/MainFrame.java:49]: Same problem

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
+package billgates.view.gui;
+
+import javax.swing.*;
+import java.awt.*;
+
+public class SettingDialog extends JDialog {
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/SettingDialog.java:6]: image
Doesn't work well on windows

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        fontField.addItem("Copperplate");
+
+
+        // Case 1: Click cancelButton
+        cancelButton.addActionListener(e -> {
+            setVisible(false);
+        });
+
+        // Case 2: Choose different color and font
+        confirmButton.addActionListener((e ->
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/SettingDialog.java:65]: In this case, we should set `colorField` and `fontField` show whatever color and font that being chosen, it shouldn't be "Default" after we change the color or font.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 10]
```diff
@@ -26,17 +34,98 @@ private void initMenu() {
...
-        this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
+        this.add(this.settingsMenu);
+
         // importMenu should be disabled  before signing in
-        this.importMenu.setEnabled(false);
+//        this.importMenu.setEnabled(false);
+
+        // Add function to import menu
+        importMenu.addMenuListener(new MenuListener() {
+            @Override
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:44]: Maybe we could make a `MenuAdapter` here, so that we don't have to override the methods that we don't need each time.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 11]
```diff
@@ -26,17 +34,98 @@ private void initMenu() {
...
+
+            @Override
+            public void menuCanceled(MenuEvent e) {
+
+            }
+        });
+        this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
+
+        // Add function to setting menu
+        settingsMenu.addMenuListener(new MenuListener() {
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:62]: Since the layout of the pop dialog is different in macOS and Windows, we can make the menu items `JCheckBoxMenuItem` (instead of displaying a dialog). And I think this will solve most of the problems that I mentioned in the previous review. 

For example,
image

However, you could still use your own design, but you have to fix the bugs that I mentioned.

**Review by CharlotteChenyy — CHANGES_REQUESTED** [review]
Yesterday I just took a quick look at your code. 

Today I went through it in detail. And I found some problems and wrote some more detailed solutions.

File: src/main/java/billgates/view/gui/TopMenuBar.java (hunk 10 above)
Comment (eevadai) [reply src/main/java/billgates/view/gui/TopMenuBar.java:44]: Fix it!

File: src/main/java/billgates/view/gui/ActionPanel.java [hunk 12]
```diff
@@ -18,9 +18,9 @@ public class ActionPanel extends JPanel {
     public static final int DEFAULT_SIGN_IN_PANEL_HEIGHT = DEFAULT_HEIGHT / 7;
     public static final int HORIZONTAL_GAP = 5;
     public static final int VERTICAL_GAP = 10;
-//    public static final int BORDER_THICKNESS = 3;
-//    public static final int EMPTY_BORDER_THICKNESS = 7;
-//    public static final Color DEFAULT_BORDER_TEXT_COLOR = new Color(220, 120, 150);
+    public static final int BORDER_THICKNESS = 3;
+    public static final int EMPTY_BORDER_THICKNESS = 7;
+    public static final Color DEFAULT_BORDER_TEXT_COLOR = new Color(220, 120, 150);
```
[the remark this answers is not in the document]
Comment (eevadai) [reply src/main/java/billgates/view/gui/ActionPanel.java:23]: Fix it!

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 13]
```diff
@@ -26,17 +34,98 @@ private void initMenu() {
         this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
 
         this.add(this.importMenu);
-        this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
+        this.add(this.settingsMenu);
```
[the remark this answers is not in the document]
Comment (eevadai) [reply src/main/java/billgates/view/gui/TopMenuBar.java:37]: Fix it!

File: src/main/java/billgates/view/gui/CustomTitleBorder.java (hunk 3 above)
Comment (eevadai) [reply src/main/java/billgates/view/gui/CustomTitleBorder.java:14]: Fix it!

File: src/main/java/billgates/view/gui/MainFrame.java (hunk 4 above)
Comment (eevadai) [reply src/main/java/billgates/view/gui/MainFrame.java:8]: Fix it

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 14]
```diff
@@ -34,69 +36,47 @@ private void initMenu() {
...
     private int[] setting() {
-        SettingDialog setdlg = new SettingDialog(this);
+//        SettingDialog setdlg = mainFrame.getSettingDialog();
+        SettingDialog setdlg = new SettingDialog(mainFrame);
         int[] setting = new int[2];
         if (setdlg.exec()) {
             int colorTargetIndex = setdlg.getMyColor();
             int fontTargetIndex = setdlg.getMyFont();
+            SettingDialog.setFontField(fontTargetIndex);
+            SettingDialog.setColorField(colorTargetIndex);
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:79]: Seems like you have some problems with the action logic here.
These two lines shouldn't be there

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 15]
```diff
@@ -114,9 +94,10 @@ public JMenu getImportMenu() {
     private void importBills() {
         System.out.println("imported");
         JFileChooser chooser = new JFileChooser();
-        FileNameExtensionFilter filter = new FileNameExtensionFilter("Your Bills", "csv", "xlsx");
+        //JDialog chooserDialog = chooser.createDialog(mainFrame);
+        FileNameExtensionFilter filter = new FileNameExtensionFilter("Your Bills", "csv");
         chooser.setFileFilter(filter);
-        int ret = chooser.showOpenDialog(this);
+        int ret = chooser.showOpenDialog(mainFrame);
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:100]: Good try to make the dialog in the center of the main window!
But there's a better way of doing this.
Later, I can commit to your branch to help you with it.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 16]
```diff
@@ -128,4 +109,17 @@ public void loadFile(File file) {
         // CONTINUE!!!
         System.out.println("loaded");
     }
-}
+
+    // Create a MenuAdapter to simplify our code
+    public abstract class MenuAdaptor implements MenuListener{
+        protected MenuAdaptor() {}
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:115]: You could make `MenuAdaptor` a separate class in a new `.java` file.

**Review by CharlotteChenyy — CHANGES_REQUESTED** [review]
Great! I can see your efforts and progress in this new commit 👍👍👍.
But there are still some problems with your code.
And I don't know why the format of your code seems a little bit weird.

**Comment by CharlotteChenyy** [comment]
In addition to the review above, there are another two problems:

1.  As ray mentioned, there's a problem with your `settingsMenu`'s state

> After closing the settings menu either using the "X", "Confirm", or "Cancel", it seems like the menu bar's setting is still being pressed, hence, I cannot click on settings again. In order to open the settings menu again, I would need to click elsewhere aside from the settings button to reset the state of the settings button, then I can press on it again.

2. There's a problem with your `Cancel` button. For example, assume that I have already set a different color and font for our GUI first. And then, if I click the `Cancel` button when I try to reset it to another color, our GUI will immediately back to its default settings.

The video below shows both of the problems:

https://user-images.githubusercontent.com/60980117/203691494-f6749888-b376-4008-b372-520614f8a98c.mp4

File: src/main/java/billgates/view/gui/MainFrame.java [hunk 17]
```diff
@@ -46,11 +47,12 @@ public MainFrame() {
         this.add(this.billPanel, BorderLayout.CENTER);
     }
 
-    public BillPanel getBillPanel() {
-        return this.billPanel;
+    public  BillPanel getBillPanel() {
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/MainFrame.java:50]: Why did you add another space here?

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 18]
```diff
@@ -1,20 +1,30 @@
...
 import javax.swing.*;
+import javax.swing.event.MenuEvent;
+import javax.swing.event.MenuListener;
+import javax.swing.filechooser.FileNameExtensionFilter;
 import java.awt.*;
+import java.awt.event.KeyEvent;
+import java.io.File;
 
-public class TopMenuBar extends JMenuBar {
+public class TopMenuBar extends JMenuBar{
```
Comment (CharlotteChenyy) [inline src/main/java/billgates/view/gui/TopMenuBar.java:11]: There should be a space before `{`

File: .gitignore [hunk 19]
```diff
@@ -120,4 +120,4 @@ gradle-app.setting
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
-replay_pid*
\ No newline at end of file
+replay_pid*.DS_Store
```
Comment (ScottCTD) [inline .gitignore:123]: I believe the `.DS_Store` should be on the next line.

File: src/main/java/billgates/view/gui/ActionPanel.java [hunk 20]
```diff
@@ -223,4 +220,25 @@ private void deleteEntry() {
...
         return deleteEntryButton;
     }
-}
+
+    // Change color in ActionPanel and statistic text area
+    public void changeColor(Color c){
+        this.setBorder(new CustomTitleBorder("Action", c));
+        this.statisticsTextArea.setBorder(new CustomTitleBorder("Statistics", c));
+        this.statisticsTextArea.setForeground(c);
+    }
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/ActionPanel.java:229]: `changeColor` should also change the color of the buttons.

You can simply add `ActionButton.allButton.forEach(b -> b.setBackground(c));` to achieve this.

File: src/main/java/billgates/view/gui/BillPanel.java [hunk 21]
```diff
@@ -55,4 +51,16 @@ private void billClick() {
     public JTable getBillTable() {
         return this.billTable;
     }
-}
+
+    // Change color of table header and border in bill panel
+    public void changeColor(Color c){
+        billTable.getTableHeader().setBackground(c);
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/BillPanel.java:57]: Convention: `this.` should be added in front of the instance variable `billTable`.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 22]
```diff
@@ -1,20 +1,30 @@
...
 
     public static final int DEFAULT_HEIGHT = 23;
     private final BoxLayout layout = new BoxLayout(this, BoxLayout.X_AXIS);
     private final JMenu importMenu = new GeneralMenu("Import");
     private final JMenu settingsMenu = new GeneralMenu("Settings");
     private final JMenu helpMenu = new GeneralMenu("Help");
+    private final Color[] colors = new Color[]{new Color(240, 140, 170), Color.RED, Color.ORANGE,
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:18]: It looks like these are constants, so you can convert them to `static` variables. Also, don't forget to update their variable names to adhere constants naming convention: `COLORS`.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 23]
```diff
@@ -1,20 +1,30 @@
...
+            Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.PINK, Color.WHITE};
+    private final String[] fonts = new String[]{"Modern No. 20", "Times New Roman", "Helvetica", "Arial", "Impact",
+            "Verdana", "Century", "Tahoma", "Copperplate"};
+    private MainFrame mainFrame;
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:22]: It seems like this is a `final` variable. The `mainFrame` instance would never change in the lifetime of our software.

File: src/main/java/billgates/view/gui/TopMenuBar.java (hunk 18 above)
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:11]: Convention: There should be a space after `extends JMenuBar`.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 24]
```diff
@@ -27,16 +37,89 @@ private void initMenu() {
...
-        // importMenu should be disabled  before signing in
-        this.importMenu.setEnabled(false);
-
         this.add(this.settingsMenu);
         this.add(Box.createRigidArea(new Dimension(ActionPanel.VERTICAL_GAP, 0)));
-
         this.add(this.helpMenu);
+
+        // Add function to import menu
+        importMenu.addMenuListener(new MenuAdaptor(){
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:45]: I don't think you are doing the right way of designing the menu.

I suggest you read https://docs.oracle.com/javase/tutorial/uiswing/components/menu.html first and then go back recode this part.

For example, on the menu bar, there should be a File menu. The import bill `MenuItem` should be in the File menu, instead of laying out directly in the menu bar.

Similarly, there should be a Setting menu in the menubar. And the "change color and font" `MenuItem`s should be in the Setting menu. BTW, it deserves a better name like "Appearance".

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 25]
```diff
@@ -27,16 +37,89 @@ private void initMenu() {
...
+    }
+
+    //Function to load file after loading file
+    public void loadFile(File file) {
+        // CONTINUE!!!
+        System.out.println("loaded");
+    }
+
+    // Create a MenuAdapter to simplify our code
+    public abstract class MenuAdaptor implements MenuListener{
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:114]: This may not be the proper way to handle menu events as I mentioned above.

You may check that documentation and then recode the event handling part.

**Review by ScottCTD — CHANGES_REQUESTED** [review]

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+        this.setTitle("Setting");
+        this.setModal(true);
+        this.setSize(400, 200);
+        this.setResizable(false);
+        SwingUtil.centerInOwner(this, this.getOwner());
+        this.setContentPane(root);
+
+        //Set the sub-action root panel
+        JPanel subAction = new JPanel();
+        subAction.add(confirmButton);
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:40]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        subAction.add(cancelButton);
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:41]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+            "Verdana", "Century", "Tahoma", "Copperplate"};
+    private static int colorIndex;
+    private static int fontIndex;
+    private boolean confirmed = false;
+    private final MainFrame mainFrame;
+
+    // Set up setting dialog to change background color and font style
+    public SettingDialog(MainFrame owner) {
+        super(owner);
+        mainFrame = owner;
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:24]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+        fontChooseBox.add(fontStyle);
+        fontChooseBox.add(fontField);
+
+        //Add the panels to root panel to set up our layout
+        root.add(colorChooseBox, BorderLayout.CENTER);
+        root.add(fontChooseBox, BorderLayout.CENTER);
+        root.add(subAction, BorderLayout.SOUTH);
+
+        // Case 1: Click cancelButton
+        cancelButton.addActionListener(e -> {
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:82]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+            setVisible(false);
+        });
+
+        // Case 2: Choose different color and font
+        confirmButton.addActionListener((e ->
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:87]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            colorIndex = colorField.getSelectedIndex();
+            colorField.setSelectedIndex(colorIndex);
+
+            fontIndex = fontField.getSelectedIndex();
+            fontField.setSelectedIndex(fontIndex);
+
+            confirmed = true;
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:95]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+
+            JOptionPane.showMessageDialog(null, "Page has been set!", "All Set", JOptionPane.PLAIN_MESSAGE);
+            setVisible(false);
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:98]: Convention: Add `this.` before using instance methods.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+
+    public static void setFontField(int index){
+        fontField.setSelectedIndex(index);
+    }
+
+    public boolean exec() {
+
+        // Show the dialog
+        this.setVisible(true);
+        return confirmed;
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:122]: Convention: Add `this.` before using instance variables.

File: src/main/java/billgates/view/gui/SettingDialog.java [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            colorIndex = colorField.getSelectedIndex();
+            colorField.setSelectedIndex(colorIndex);
+
+            fontIndex = fontField.getSelectedIndex();
+            fontField.setSelectedIndex(fontIndex);
+
+            confirmed = true;
+
+            JOptionPane.showMessageDialog(null, "Page has been set!", "All Set", JOptionPane.PLAIN_MESSAGE);
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/SettingDialog.java:97]: You can simply change `null` here to `this` and then the dialog will show in the center of the setting dialog no matter where the setting dialog currently is.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 35]
```diff
@@ -1,20 +1,29 @@
...
 
 public class TopMenuBar extends JMenuBar {
 
     public static final int DEFAULT_HEIGHT = 23;
     private final BoxLayout layout = new BoxLayout(this, BoxLayout.X_AXIS);
-    private final JMenu importMenu = new GeneralMenu("Import");
+    private final JMenu fileMenu = new GeneralMenu("File");
     private final JMenu settingsMenu = new GeneralMenu("Settings");
     private final JMenu helpMenu = new GeneralMenu("Help");
+    private final static Color[] COLORS = new Color[]{new Color(240, 140, 170), Color.RED, Color.ORANGE,
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:16]: I don't see where the two arrays are used. You may consider deleting them.

File: src/main/java/billgates/view/gui/TopMenuBar.java [hunk 36]
```diff
@@ -25,18 +34,80 @@ private void initMenu() {
...
+            int colorTargetIndex = setdlg.getMyColor();
+            int fontTargetIndex = setdlg.getMyFont();
+            SettingDialog.setFontField(fontTargetIndex);
+            SettingDialog.setColorField(colorTargetIndex);
+            setting[0] = colorTargetIndex;
+            setting[1] = fontTargetIndex;
+            return setting;
+        }
+        setting[0] = settingDialog.getMyColor();
+        setting[1] = settingDialog.getMyFont();
```
Comment (ScottCTD) [inline src/main/java/billgates/view/gui/TopMenuBar.java:86]: These two lines cause a `NullPointerException` when you click the cancel button.
This is because your `settingDialog` does not initialize.
The simple solution is change `settingDialog` to `setdlg`.

Also, remember to remove the `settingDialog` instance variable declaration: `settingDialog` are useless (at least I don't see anywhere you used it).

**Review by ScottCTD — CHANGES_REQUESTED** [review]
